File: agenkit-java/src/main/java/io/agenkit/observability/MetricsCollector.java
package io.agenkit.observability;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Collects agent metrics for monitoring.
 */
public final class MetricsCollector {

    private final Map<String, AtomicLong> counters = new ConcurrentHashMap<>();
    private final Map<String, List<Double>> histograms = new ConcurrentHashMap<>();

    public void increment(String metric) {
        counters.computeIfAbsent(metric, k -> new AtomicLong(0)).incrementAndGet();
    }

    public void increment(String metric, long amount) {
        counters.computeIfAbsent(metric, k -> new AtomicLong(0)).addAndGet(amount);
    }

    public void record(String metric, double value) {
        histograms.computeIfAbsent(metric, k ->
                Collections.synchronizedList(new ArrayList<>())).add(value);
    }

    public long getCount(String metric) {
        AtomicLong counter = counters.get(metric);
        return counter != null ? counter.get() : 0;
    }

    public double getAverage(String metric) {
        List<Double> values = histograms.get(metric);
        if (values == null || values.isEmpty()) return 0.0;
        synchronized (values) {
            return values.stream().mapToDouble(Double::doubleValue).average().orElse(0.0);
        }
    }

    public Map<String, Long> getAllCounters() {
        Map<String, Long> result = new ConcurrentHashMap<>();
        counters.forEach((k, v) -> result.put(k, v.get()));
        return Collections.unmodifiableMap(result);
    }
}

File: agenkit-scala/src/main/scala/io/agenkit/observability/MetricsCollector.scala
package io.agenkit.observability

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong
import scala.jdk.CollectionConverters.*

/** Thread-safe metrics collector with counters, gauges, and histograms. */
class MetricsCollector:
  private val counters   = new ConcurrentHashMap[String, AtomicLong]()
  private val gauges     = new ConcurrentHashMap[String, AtomicLong]()
  private val histograms = new ConcurrentHashMap[String, collection.mutable.ListBuffer[Double]]()

  def incrementCounter(name: String, value: Long = 1): Unit =
    counters.computeIfAbsent(name, _ => new AtomicLong(0)).addAndGet(value)
    ()

  def setGauge(name: String, value: Long): Unit =
    gauges.computeIfAbsent(name, _ => new AtomicLong(0)).set(value)

  def recordHistogram(name: String, value: Double): Unit =
    val buf = histograms.computeIfAbsent(name, _ => collection.mutable.ListBuffer[Double]())
    buf.synchronized { buf.addOne(value) }

  def getCounter(name: String): Long =
    Option(counters.get(name)).map(_.get()).getOrElse(0L)

  def getGauge(name: String): Long =
    Option(gauges.get(name)).map(_.get()).getOrElse(0L)

  def getHistogramStats(name: String): Map[String, Double] =
    Option(histograms.get(name)).map { buf =>
      val values = buf.synchronized { buf.toList }
      if values.isEmpty then Map("count" -> 0.0, "mean" -> 0.0, "p99" -> 0.0)
      else
        val sorted = values.sorted
        Map(
          "count" -> values.size.toDouble,
          "mean"  -> values.sum / values.size,
          "p50"   -> sorted(values.size / 2),
          "p99"   -> sorted((values.size * 0.99).toInt.min(values.size - 1))
        )
    }.getOrElse(Map.empty)

  def snapshot(): Map[String, Any] =
    Map(
      "counters" -> counters.keys().asScala.toList.map(k => k -> getCounter(k)).toMap,
      "gauges"   -> gauges.keys().asScala.toList.map(k => k -> getGauge(k)).toMap
    )
